File: frontend/src/services/weatherService.ts
/**
 * weatherService.ts
 *
 * Calls Open-Meteo and its Geocoding API **directly from the browser**.
 * No backend is required — the weather page works even when the FastAPI
 * server is offline.
 */

import { WeatherData, ForecastDay } from '../types/weather';

// ─── WMO Weather Code map ─────────────────────────────────────────────────────
const WMO: Record<number, [string, string]> = {
  0:  ['Clear Sky',                   'sun'],
  1:  ['Mainly Clear',                'cloud-sun'],
  2:  ['Partly Cloudy',               'cloud-sun'],
  3:  ['Overcast',                    'cloud'],
  45: ['Fog',                         'cloud-fog'],
  48: ['Depositing Rime Fog',         'cloud-fog'],
  51: ['Light Drizzle',               'cloud-drizzle'],
  53: ['Moderate Drizzle',            'cloud-drizzle'],
  55: ['Dense Drizzle',               'cloud-drizzle'],
  61: ['Slight Rain',                 'cloud-rain'],
  63: ['Moderate Rain',               'cloud-rain'],
  65: ['Heavy Rain',                  'cloud-rain'],
  71: ['Slight Snow',                 'snowflake'],
  73: ['Moderate Snow',               'snowflake'],
  75: ['Heavy Snow',                  'snowflake'],
  80: ['Rain Showers',                'cloud-rain'],
  81: ['Moderate Rain Showers',       'cloud-rain'],
  82: ['Violent Rain Showers',        'cloud-rain'],
  85: ['Slight Snow Showers',         'snowflake'],
  86: ['Heavy Snow Showers',          'snowflake'],
  95: ['Thunderstorm',                'cloud-lightning'],
  96: ['Thunderstorm w/ Slight Hail', 'cloud-lightning'],
  99: ['Thunderstorm w/ Heavy Hail',  'cloud-lightning'],
};

const DAY_ABBR = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

function parseWmo(code: number | null | undefined): [string, string] {
  if (code == null) return ['Partly Cloudy', 'cloud-sun'];
  return WMO[code] ?? ['Partly Cloudy', 'cloud-sun'];
}

function buildInsight(temp: number, humidity: number, rainProb: number, precip: number, soilTemp: number): string {
  if (rainProb >= 60 || precip > 2.0) {
    return `High precipitation likelihood (${rainProb.toFixed(0)}%). Postpone irrigation cycles and foliar spraying.`;
  }
  if (rainProb >= 30) {
    return `Scattered rain possible (${rainProb.toFixed(0)}%). Check rootzone moisture before field irrigation.`;
  }
  if (temp >= 36) {
    return `Elevated temperature (${temp.toFixed(1)}°C) with high evapotranspiration demand. Ensure adequate soil hydration.`;
  }
  if (humidity >= 85) {
    return `High relative humidity (${humidity.toFixed(0)}%). Monitor canopy for foliar fungal and bacterial spore proliferation.`;
  }
  return `Optimal agro-climatic conditions. Soil temperature is ${soilTemp.toFixed(1)}°C — suitable for paddy maintenance.`;
}

// ─── Geocoding ────────────────────────────────────────────────────────────────
async function geocodeCity(city: string): Promise<{ lat: number; lng: number; name: string } | null> {
  const url = `https://geocoding-api.open-meteo.com/v1/search?name=${encodeURIComponent(city.trim())}&count=1&language=en&format=json`;
  const res = await fetch(url);
  if (!res.ok) return null;
  const json = await res.json();
  const results: any[] = json.results ?? [];
  if (!results.length) return null;
  const top = results[0];
  const parts = [top.name, top.admin1, top.country].filter(Boolean);
  return {
    lat: parseFloat(top.latitude),
    lng: parseFloat(top.longitude),
    name: parts.join(', '),
  };
}

// ─── Open-Meteo forecast fetch ────────────────────────────────────────────────
async function fetchOpenMeteo(lat: number, lng: number, locationName: string): Promise<WeatherData> {
  const url =
    `https://api.open-meteo.com/v1/forecast?` +
    `latitude=${lat}&longitude=${lng}` +
    `&current=temperature_2m,relative_humidity_2m,precipitation,precipitation_probability,wind_speed_10m,weather_code,soil_temperature_0cm` +
    `&daily=temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,relative_humidity_2m_mean,uv_index_max,weather_code` +
    `&timezone=auto`;

  const res = await fetch(url);
  if (!res.ok) throw new Error(`Open-Meteo responded with status ${res.status}`);
  const data = await res.json();

  const cur = data.current ?? {};
  const daily = data.daily ?? {};

  const currTemp     = parseFloat(cur.temperature_2m      ?? 28);
  const currHumidity = parseFloat(cur.relative_humidity_2m ?? 70);
  const currRainProb = parseFloat(cur.precipitation_probability ?? 0) || 0;
  const currWind     = parseFloat(cur.wind_speed_10m       ?? 10);
  const currPrecip   = parseFloat(cur.precipitation        ?? 0) || 0;
  const currSoilTemp = parseFloat(cur.soil_temperature_0cm ?? (currTemp - 2));
  const [currCondition] = parseWmo(cur.weather_code);

  const dates   = (daily.time                        ?? []) as string[];
  const tMax    = (daily.temperature_2m_max          ?? []) as number[];
  const tMin    = (daily.temperature_2m_min          ?? []) as number[];
  const pSum    = (daily.precipitation_sum           ?? []) as number[];
  const pProb   = (daily.precipitation_probability_max ?? []) as number[];
  const hum     = (daily.relative_humidity_2m_mean   ?? []) as number[];
  const uvMax   = (daily.uv_index_max                ?? []) as number[];
  const codes   = (daily.weather_code                ?? []) as number[];

  const forecast: ForecastDay[] = dates.slice(0, 7).map((d, i) => {
    const date = new Date(d);
    const dayLabel = i === 0 ? 'Today' : DAY_ABBR[date.getDay()];
    const [condition, icon] = parseWmo(codes[i]);
    return {
      date: d,
      dayName: dayLabel,
      tempMax: Math.round((tMax[i] ?? 30) * 10) / 10,
      tempMin: Math.round((tMin[i] ?? 22) * 10) / 10,
      condition,
      rainProbability: Math.round(pProb[i] ?? 0),
      humidity: Math.round(hum[i] ?? 65),
      icon,
      precipitationMm: Math.round((pSum[i] ?? 0) * 10) / 10,
      uvIndex: Math.round((uvMax[i] ?? 5) * 10) / 10,
    };
  });

  const todayUv = uvMax[0] ?? 5;

  return {
    location: locationName,
    latitude: Math.round(lat * 10000) / 10000,
    longitude: Math.round(lng * 10000) / 10000,
    currentTemp: Math.round(currTemp * 10) / 10,
    condition: currCondition,
    humidity: Math.round(currHumidity * 10) / 10,
    rainProbability: Math.round(currRainProb),
    windSpeedKmH: Math.round(currWind * 10) / 10,
    uvIndex: Math.round(todayUv * 10) / 10,
    soilTemp: Math.round(currSoilTemp * 10) / 10,
    precipitationMm: Math.round(currPrecip * 10) / 10,
    forecast,
    farmInsight: buildInsight(currTemp, currHumidity, currRainProb, currPrecip, currSoilTemp),
    lastUpdated: 'Just now',
    source: 'open-meteo',
  };
}

// ─── Public API ───────────────────────────────────────────────────────────────
export const weatherService = {
  /** Fetch weather by GPS coordinates directly from Open-Meteo. */
  async getWeatherByCoordinates(latitude: number, longitude: number): Promise<WeatherData> {
    const locationName = `${latitude.toFixed(4)}° N, ${longitude.toFixed(4)}° E`;
    try {
      return await fetchOpenMeteo(latitude, longitude, locationName);
    } catch (err: any) {
      throw new Error('Unable to fetch live weather. Please check your internet connection and try again.');
    }
  },

  /** Geocode city name then fetch weather from Open-Meteo. */
  async getWeatherByCity(city: string): Promise<WeatherData> {
    const trimmed = city.trim();
    if (!trimmed) throw new Error('Please enter a city or village name.');

    // Try geocoding first
    let lat = 22.3072;
    let lng = 73.1812;
    let name = trimmed;

    try {
      const geo = await geocodeCity(trimmed);
      if (geo) {
        lat = geo.lat;
        lng = geo.lng;
        name = geo.name;
      } else {
        throw new Error(`Location '${trimmed}' not found. Please check the spelling and try again.`);
      }
    } catch (geoErr: any) {
      // If the error is "not found", rethrow. Otherwise network error — try with defaults.
      if (geoErr.message?.includes('not found')) throw geoErr;
      // Network issue during geocoding — use default Vadodara coords as fallback
      name = trimmed;
    }

    try {
      return await fetchOpenMeteo(lat, lng, name);
    } catch (err: any) {
      throw new Error('Unable to fetch live weather. Please check your internet connection and try again.');
    }
  },

  /** Generic fetcher supporting coordinates or city search. */
  async getWeather(params: { latitude?: number; longitude?: number; city?: string }): Promise<WeatherData> {
    if (params.latitude !== undefined && params.longitude !== undefined) {
      return this.getWeatherByCoordinates(params.latitude, params.longitude);
    }
    if (params.city) {
      return this.getWeatherByCity(params.city);
    }
    throw new Error('Either coordinates (latitude & longitude) or a city name is required.');
  },
};

File: backend/app/services/weather_service.py
import datetime
import httpx
from typing import Optional, List, Dict, Any, Tuple
from app.core.config import settings
from app.core.logging import logger
from app.schemas.weather import WeatherDataResponse, ForecastDay

# WMO Weather interpretation code (WW) mapping to human condition and icon
WMO_WEATHER_CODES: Dict[int, Tuple[str, str]] = {
    0: ("Clear Sky", "sun"),
    1: ("Mainly Clear", "cloud-sun"),
    2: ("Partly Cloudy", "cloud-sun"),
    3: ("Overcast", "cloud"),
    45: ("Fog", "cloud-fog"),
    48: ("Depositing Rime Fog", "cloud-fog"),
    51: ("Light Drizzle", "cloud-drizzle"),
    53: ("Moderate Drizzle", "cloud-drizzle"),
    55: ("Dense Drizzle", "cloud-drizzle"),
    56: ("Light Freezing Drizzle", "cloud-drizzle"),
    57: ("Dense Freezing Drizzle", "cloud-drizzle"),
    61: ("Slight Rain", "cloud-rain"),
    63: ("Moderate Rain", "cloud-rain"),
    65: ("Heavy Rain", "cloud-rain"),
    66: ("Light Freezing Rain", "cloud-rain"),
    67: ("Heavy Freezing Rain", "cloud-rain"),
    71: ("Slight Snow", "snowflake"),
    73: ("Moderate Snow", "snowflake"),
    75: ("Heavy Snow", "snowflake"),
    77: ("Snow Grains", "snowflake"),
    80: ("Slight Rain Showers", "cloud-rain"),
    81: ("Moderate Rain Showers", "cloud-rain"),
    82: ("Violent Rain Showers", "cloud-rain"),
    85: ("Slight Snow Showers", "snowflake"),
    86: ("Heavy Snow Showers", "snowflake"),
    95: ("Thunderstorm", "cloud-lightning"),
    96: ("Thunderstorm with Slight Hail", "cloud-lightning"),
    99: ("Thunderstorm with Heavy Hail", "cloud-lightning"),
}


def parse_wmo_code(code: Optional[int]) -> Tuple[str, str]:
    """Return condition text and icon string for a given WMO weather code."""
    if code is None:
        return "Partly Cloudy", "cloud-sun"
    return WMO_WEATHER_CODES.get(int(code), ("Partly Cloudy", "cloud-sun"))


async def geocode_city_name(city: str) -> Optional[Tuple[float, float, str]]:
    """
    Search Open-Meteo Geocoding API for city/village.
    Returns (latitude, longitude, formatted_location_string) or None if not found.
    """
    url = f"https://geocoding-api.open-meteo.com/v1/search?name={city.strip()}&count=1&language=en&format=json"
    try:
        async with httpx.AsyncClient(timeout=4.0) as client:
            res = await client.get(url)
            if res.status_code == 200:
                data = res.json()
                results = data.get("results")
                if results and len(results) > 0:
                    top = results[0]
                    lat = float(top["latitude"])
                    lng = float(top["longitude"])
                    name = top.get("name", city)
                    admin1 = top.get("admin1")
                    country = top.get("country")
                    parts = [p for p in [name, admin1, country] if p]
                    formatted_name = ", ".join(parts)
                    return lat, lng, formatted_name
    except Exception as e:
        logger.warning(f"Open-Meteo geocoding request failed for '{city}': {e}")
    return None


class WeatherProvider:
    async def get_weather(
        self,
        location: str,
        lat: Optional[float] = None,
        lng: Optional[float] = None
    ) -> WeatherDataResponse:
        raise NotImplementedError


class MockWeatherProvider(WeatherProvider):
    async def get_weather(
        self,
        location: str,
        lat: Optional[float] = None,
        lng: Optional[float] = None
    ) -> WeatherDataResponse:
        day_names = ["Today", "Tomorrow", "Wed", "Thu", "Fri", "Sat", "Sun"]
        today = datetime.date.today()

        forecast = [
            ForecastDay(
                date=(today + datetime.timedelta(days=i)).isoformat(),
                dayName=day_names[i] if i < len(day_names) else f"Day {i+1}",
                tempMax=31.0 + (i % 3),
                tempMin=22.0 + (i % 2),
                condition="Partly Cloudy" if i % 2 == 0 else "Scattered Showers",
                rainProbability=30.0 + (i * 5),
                humidity=70.0 + (i * 2),
                icon="cloud-sun" if i % 2 == 0 else "cloud-rain",
                precipitationMm=0.0 if i % 2 == 0 else 2.5,
                uvIndex=6.0
            )
            for i in range(7)
        ]

        return WeatherDataResponse(
            location=location or "Vadodara, Gujarat, India",
            latitude=lat or 22.3072,
            longitude=lng or 73.1812,
            currentTemp=28.0,
            condition="Partly Cloudy",
            humidity=72.0,
            rainProbability=30.0,
            windSpeedKmH=14.0,
            uvIndex=6.0,
            soilTemp=24.0,
            precipitationMm=0.0,
            forecast=forecast,
            farmInsight="Simulated fallback telemetry. Open-Meteo service was unreachable; please verify internet connectivity.",
            lastUpdated="Just now",
            source="fallback"
        )


class OpenMeteoWeatherProvider(WeatherProvider):
    async def get_weather(
        self,
        location: str,
        lat: Optional[float] = None,
        lng: Optional[float] = None
    ) -> WeatherDataResponse:
        latitude = lat if lat is not None else 22.3072
        longitude = lng if lng is not None else 73.1812

        url = (
            f"https://api.open-meteo.com/v1/forecast?"
            f"latitude={latitude}&longitude={longitude}"
            f"&current=temperature_2m,relative_humidity_2m,precipitation,precipitation_probability,wind_speed_10m,weather_code,soil_temperature_0cm"
            f"&daily=temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,relative_humidity_2m_mean,uv_index_max,weather_code"
            f"&timezone=auto"
        )

        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                res = await client.get(url)
                if res.status_code == 200:
                    data = res.json()
                    current = data.get("current", {})
                    daily = data.get("daily", {})

                    # Extract current weather variables
                    curr_temp = float(current.get("temperature_2m", 28.0))
                    curr_humidity = float(current.get("relative_humidity_2m", 70.0))
                    curr_rain_prob = float(current.get("precipitation_probability", 0.0) or 0.0)
                    curr_wind = float(current.get("wind_speed_10m", 10.0))
                    curr_precip_mm = float(current.get("precipitation", 0.0) or 0.0)
                    curr_soil_temp = float(current.get("soil_temperature_0cm", curr_temp - 2.0))
                    curr_code = current.get("weather_code", 0)
                    curr_condition, _ = parse_wmo_code(curr_code)

                    # Extract 7-day daily forecast variables
                    dates = daily.get("time", [])
                    t_max = daily.get("temperature_2m_max", [])
                    t_min = daily.get("temperature_2m_min", [])
                    p_sum = daily.get("precipitation_sum", [])
                    p_prob = daily.get("precipitation_probability_max", [])
                    hum = daily.get("relative_humidity_2m_mean", [])
                    uv_max = daily.get("uv_index_max", [])
                    codes = daily.get("weather_code", [])

                    forecast_days: List[ForecastDay] = []
                    day_names_abbr = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

                    num_days = min(7, len(dates))
                    for i in range(num_days):
                        d_str = dates[i]
                        try:
                            parsed_date = datetime.date.fromisoformat(d_str)
                            weekday_name = day_names_abbr[parsed_date.weekday()]
                        except Exception:
                            weekday_name = f"Day {i+1}"

                        day_label = "Today" if i == 0 else weekday_name
                        code_val = codes[i] if i < len(codes) else 0
                        cond_name, icon_name = parse_wmo_code(code_val)

                        forecast_days.append(
                            ForecastDay(
                                date=d_str,
                                dayName=day_label,
                                tempMax=round(float(t_max[i]), 1) if i < len(t_max) and t_max[i] is not None else 30.0,
                                tempMin=round(float(t_min[i]), 1) if i < len(t_min) and t_min[i] is not None else 22.0,
                                condition=cond_name,
                                rainProbability=round(float(p_prob[i]), 0) if i < len(p_prob) and p_prob[i] is not None else 0.0,
                                humidity=round(float(hum[i]), 0) if i < len(hum) and hum[i] is not None else 65.0,
                                icon=icon_name,
                                precipitationMm=round(float(p_sum[i]), 1) if i < len(p_sum) and p_sum[i] is not None else 0.0,
                                uvIndex=round(float(uv_max[i]), 1) if i < len(uv_max) and uv_max[i] is not None else 5.0
                            )
                        )

                    # Determine today's UV index from daily UV max
                    today_uv = float(uv_max[0]) if uv_max and len(uv_max) > 0 and uv_max[0] is not None else 5.0

                    # Dynamic agricultural farm insight
                    if curr_rain_prob >= 60 or curr_precip_mm > 2.0:
                        insight = f"High precipitation likelihood ({curr_rain_prob:.0f}%). Postpone scheduled irrigation cycles and foliar fungicide spraying."
                    elif curr_rain_prob >= 30:
                        insight = f"Scattered rain possible ({curr_rain_prob:.0f}%). Inspect soil rootzone moisture before initiating field irrigation."
                    elif curr_temp >= 36:
                        insight = f"Elevated temperature ({curr_temp:.1f}°C) with high evapotranspiration demand. Ensure adequate soil hydration."
                    elif curr_humidity >= 85:
                        insight = f"High ambient relative humidity ({curr_humidity:.0f}%). Monitor canopy closely for foliar fungal and bacterial spore proliferation."
                    else:
                        insight = f"Optimal agro-climatic conditions for paddy maintenance. Soil temperature is {curr_soil_temp:.1f}°C."

                    return WeatherDataResponse(
                        location=location,
                        latitude=round(latitude, 4),
                        longitude=round(longitude, 4),
                        currentTemp=round(curr_temp, 1),
                        condition=curr_condition,
                        humidity=round(curr_humidity, 1),
                        rainProbability=round(curr_rain_prob, 0),
                        windSpeedKmH=round(curr_wind, 1),
                        uvIndex=round(today_uv, 1),
                        soilTemp=round(curr_soil_temp, 1),
                        precipitationMm=round(curr_precip_mm, 1),
                        forecast=forecast_days,
                        farmInsight=insight,
                        lastUpdated="Just now",
                        source="open-meteo"
                    )
        except Exception as e:
            logger.warning(f"Open-Meteo live query failed: {e}. Returning fallback telemetry.")

        return await MockWeatherProvider().get_weather(location, lat, lng)


class WeatherService:
    def __init__(self):
        self.live_provider: WeatherProvider = OpenMeteoWeatherProvider()
        self.mock_provider: WeatherProvider = MockWeatherProvider()

    async def get_current_weather(
        self,
        location: Optional[str] = None,
        lat: Optional[float] = None,
        lng: Optional[float] = None
    ) -> WeatherDataResponse:
        return await self.live_provider.get_weather(location or "Vadodara, Gujarat, India", lat, lng)


weather_service = WeatherService()
